src/security/builtin_plugins/authentication/src/authentication.c

```diff
@@ -1429,38 +1360,34 @@ validate_remote_identity(
      */
 
     /* Check if the remote identity has already been validated by a previous validation request. */
-    remoteIdent = find_remote_identity_by_guid(impl->remoteGuidHash, remote_participant_guid);
-    if (!remoteIdent) {
-        remoteIdent = remoteIdentityInfoNew(remote_participant_guid, remote_identity_token);
-        (void)ddsrt_hh_add(impl->objectHash, remoteIdent);
-        (void)ddsrt_hh_add(impl->remoteGuidHash, remoteIdent);
-        relation = identityRelationNew(localIdent, remoteIdent, lchallenge, rchallenge);
-        (void)ddsrt_hh_add(remoteIdent->linkHash, relation);
+    rinfo = find_remote_identity_by_guid(impl->remoteGuidHash, remote_participant_guid);
+    if (!rinfo) {
+        rinfo = remoteIdentityInfoNew(remote_participant_guid, remote_identity_token);
+        (void)ddsrt_hh_add(impl->remoteGuidHash, rinfo);
     } else {
         /* When the remote identity has already been validated before,
            check if the remote identity token matches with the existing one
          */
-        if (!DDS_Security_DataHolder_equal(remoteIdent->remoteIdentityToken, remote_identity_token)) {
+        if (!DDS_Security_DataHolder_equal(rinfo->remoteIdentityToken, remote_identity_token)) {
             result = DDS_SECURITY_VALIDATION_FAILED;
             DDS_Security_Exception_set(ex, DDS_AUTH_PLUGIN_CONTEXT, DDS_SECURITY_ERR_UNDEFINED_CODE, (int)result,
                     "validate_remote_identity: remote_identity_token does not match with previously received one");
             goto err_inv_duplicate;
         }
+        relation = ddsrt_avl_lookup(&relations_treedef, &linfo->relations, rinfo);
```

Comment:
While it is fine to use ``ddsrt_avl_lookup`` followed by ``ddsrt_avl_insert`` if it isn't found, it does effectively require a second lookup inside the ``ddsrt_avl_insert``. There is a version suffixed with ``_ipath`` that allows you to avoid the second lookup by caching information in the lookup.